hail/python/hail/methods/relatedness/pc_air.py

```diff
@@ -0,0 +1,278 @@
+    We say that two samples are **divergent** if their ancestral divergence is less than
+    the negation of the divergence threshold.
+
+    This method estimates the kinship coefficient and the ancestral divergence between all samples
+    using the `KING-robust, between-family kinship coefficient <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3025716/>`_
+    estimator.
+
+    This method returns an unrelated set and a related set.
+    The intersection of these sets is empty, and the union of these sets is the set of all samples.
+    Thus, the unrelated set and the related set are a **partition** of the set of all samples.
+
+    The samples in the unrelated set are mutually unrelated.
+
+    The partitioning algorithm is documented in the
+    `PC-AiR paper <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4836868/#APP2title>`_.
+
+    Parameters
+    ----------
+    genotypes : :class:`.CallExpression`
+        A call expression representing the genotype calls.
+    relatedness_threshold : :obj:`int` or :obj:`float` or :class:`.NumericExpression`
+        The relatedness threshold. The default is 0.025.
+    divergence_threshold : :obj:`int` or :obj:`float` or :class:`.NumericExpression`
+        The divergence threshold. The default is 0.025.
+
+    Returns
+    -------
+    :obj:`set` of :class:`.Struct`
+        The keys of the samples in the unrelated set.
+    :obj:`set` of :class:`.Struct`
+        The keys of the samples in the related set.
+    """
+    # The variable names in this method are based on the notation in the PC-AiR paper.
+    # TODO: The paper uses the within-family estimate for ancestral divergence
+    # TODO: The paper suggests using the within-family estimate for relatedness as well
+    # king returns the KING-robust, between-family kinship estimates for all sample pairs
+    pairs: MatrixTable = king(genotypes)
+    pairs = pairs.cache()
+
+    assert len(pairs.row_key) == len(pairs.col_key)
+    assert isinstance(pairs.row_key.dtype, hl.tstruct) and isinstance(pairs.col_key.dtype, hl.tstruct)
+    assert pairs.row_key.dtype.types == pairs.col_key.dtype.types
+
+    def keys_are_different():
+        return hl.any(
+            list(
+                pairs[left_field] != pairs[right_field]
+                for left_field, right_field in zip(pairs.row_key.dtype, pairs.col_key.dtype)
+            )
+        )
+
+    pairs = pairs.annotate_cols(eta=hl.agg.count_where(keys_are_different() & (pairs.phi > relatedness_threshold)))
+    pairs = pairs.annotate_cols(
+        delta=hl.agg.count_where(
+            keys_are_different() & (pairs.phi < relatedness_threshold) & (pairs.phi < -divergence_threshold)
+        )
+    )
+    pairs = pairs.annotate_cols(
+        gamma=hl.agg.sum(hl.if_else(keys_are_different() & (pairs.phi > relatedness_threshold), pairs.phi, 0))
+    )
+
+    unrelated = pairs.aggregate_cols(hl.agg.collect_as_set(pairs.col_key))
+    related = set()
+    samples = pairs.cols()
```

Comment:
I am starting to think that this method (`_partition_samples`) would be better if it localized the samples table and performed the algorithm locally. This method's current implementation is rather slow—it takes 63 seconds to run the unit test I wrote for this method on my machine. Even for studies with millions of samples, it seems reasonable to assume this data could fit on a single machine.

Reply:
I got the execution time of the unit test down to 37 seconds in commit 4c570be by avoiding set operations. I then made the implementation more Hail-idiomatic by using table joins instead of manually filtering and selecting the keys (commit e89f3cf). This brought the execution time up to 40 seconds, but I prefer this approach because of its simplicity.

Still wondering if `_partition_samples` should just localize the samples though...